plugins/inputs/jti_openconfig_telemetry/openconfig_telemetry.go

```diff
@@ -0,0 +1,324 @@
+	m.parser = parser
+}
+
+func (m *OpenConfigTelemetry) Gather(acc telegraf.Accumulator) error {
+	return nil
+}
+
+func (m *OpenConfigTelemetry) Stop() {
+	m.Lock()
+	defer m.Unlock()
+	m.grpcClientConn.Close()
+}
+
+// Takes in XML path with predicates and returns list of tags+values along with a final XML path without predicates
+func spitTagsNPath(xmlpath string) (string, map[string]string) {
+	re := regexp.MustCompile("\\/([^\\/]*)\\[([A-Za-z0-9\\-\\/]*)\\=([^\\[]*)\\]")
+	subs := re.FindAllStringSubmatch(xmlpath, -1)
+	tags := make(map[string]string)
+
+	// Given XML path, this will spit out final path without predicates
+	if len(subs) > 0 {
+		for _, sub := range subs {
+			tagKey := strings.Split(xmlpath, sub[0])[0]
+			tagKey += "/" + strings.TrimSpace(sub[1]) + "/@" + strings.TrimSpace(sub[2])
+			tagValue := strings.Replace(sub[3], "'", "", -1)
+
+			tags[tagKey] = tagValue
+			xmlpath = strings.Replace(xmlpath, sub[0], "/"+strings.TrimSpace(sub[1]), 1)
+		}
+	}
+
+	return xmlpath, tags
+}
+
+func (m *OpenConfigTelemetry) Start(acc telegraf.Accumulator) error {
+	log.Print("I! Started JTI OpenConfig Telemetry plugin\n")
+
+	m.Lock()
+	defer m.Unlock()
+
+	m.acc = acc
+
+	// Extract device name / IP
+	s := strings.Split(m.Server, ":")
+	grpc_server, grpc_port := s[0], s[1]
+
+	var err error
+
+	// If a certificate is provided, open a secure channel. Else open insecure one
+	if m.CertFile != "" {
+		creds, err := credentials.NewClientTLSFromFile(m.CertFile, "")
+		if err != nil {
+			log.Fatalf("E! Failed to read certificate: %v", err)
+		}
+		m.grpcClientConn, err = grpc.Dial(m.Server, grpc.WithTransportCredentials(creds))
+	} else {
+		m.grpcClientConn, err = grpc.Dial(m.Server, grpc.WithInsecure())
+	}
+	if err != nil {
+		log.Fatalf("E! Failed to connect: %v", err)
+	}
+
+	if m.Debug {
+		log.Printf("I! Opened a new gRPC session to %s on port %s", grpc_server, grpc_port)
+	}
+
+	// If username, password and clientId are provided, authenticate user before subscribing for data
+	if m.Username != "" && m.Password != "" && m.ClientId != "" {
+		lc := authentication.NewLoginClient(m.grpcClientConn)
+		loginReply, loginErr := lc.LoginCheck(context.Background(), &authentication.LoginRequest{UserName: m.Username, Password: m.Password, ClientId: m.ClientId})
+		if loginErr != nil {
+			log.Fatalf("E! Could not initiate login check: %v", err)
+		}
+
+		// Check if the user is authenticated. Bail if auth error
+		if !loginReply.Result {
+			log.Fatalf("E! Failed to authenticate the user")
+		}
+	}
+
+	c := telemetry.NewOpenConfigTelemetryClient(m.grpcClientConn)
+
+	for _, sensor := range m.Sensors {
+		go func(sensor string, acc telegraf.Accumulator) {
+			spathSplit := strings.SplitN(sensor, " ", -1)
+			var sensorName string
+			var pathlist []*telemetry.Path
+			customReportingRate := false
+			var reportingRate uint64
+
+			reportingRate = uint64(m.SampleFrequency)
+
+			if len(spathSplit) > 1 {
+				for i, path := range spathSplit {
```

Comment:
This is more complex than it needs to be, I don't think you should use a loop here.  Try assigning the components of the sensors option to variables and then take action based on their values.

Reply:
Thanks. Don't know what I was thinking before. Rewrote this.